File: kotlin/day3_2.kt
fun main() {
    val regexp = Regex("(mul\\([0-9]+,[0-9]+\\)|do\\(\\)|don't\\(\\))")
    val input = generateSequence(::readLine).joinToString("\n")
    val result = regexp.findAll(input)
    var s = 0
    var doing = true
    for (r in result) {
        val instruction = r.groups[0]!!.value
        when (instruction) {
            "do()" -> doing = true
            "don't()" -> doing = false
            else ->
                    if (doing) {
                        val ins = instruction.split(",")
                        val a = ins[0].filter { it.isDigit() }.toInt()
                        val b = ins[1].filter { it.isDigit() }.toInt()
                        s += a * b
                    }
        }
    }
    print(s)
}

File: python/day3_2.py
import sys, re


l = sys.stdin.read()
muls =re.findall("(mul\\([0-9]+,[0-9]+\\)|do\\(\\)|don't\\(\\))",l)

s =0

do = True

for m in muls:
    if m=='do()':
        do = True
    elif m=="don't()":
        do = False
    else:
        a,b=m.split(",")
        a=int(a[4:])
        b=int(b[:-1])
        if do:
            s+=a*b

print(s)
